app/logical/deferred_posts.rb

```diff
@@ -0,0 +1,31 @@
+class DeferredPosts

```

Comment:
This is a bit silly. Here's how I would probably go about this:
```rb
require 'set'

module DeferredPosts
  def deferred_post_ids
    @post_id_set ||= Set.new
  end

  def deferred_posts
    Post.where(id: deferred_post_ids.to_a).find_each.reduce({}) do |post_hash, p|
      post_hash[p.id] = p.minimal_attributes
      post_hash
    end
  end
end
```

Include the module (which for example could be in a [controller concern](https://elegantbrew.tumblr.com/post/70990048275/controller-concerns-in-rails-4)) into the base controller or any subclass thereof:
```rb
  # application_controller.rb
  include DeferredPosts
  helper_method :deferred_post_ids, :deferred_posts
```

Add deferred ids in a controller or view:
```rb
  deferred_post_ids.add(1)
```

...and dump them:
```erb
<script>
  window.__deferredPosts = <%= raw deferred_posts.to_json %>;
</script>
```

No need for request_store or anything of the sort.

Reply:
I will investigate how this interacts with dtext(the next step, and other user of this functionality) and revise this code based on that experiment. What you suggest looks very reasonable if it works for dtext.

Reply:
Will fix in the PR for dtext thumbnails.